store/secretsmanagerstore.go

```diff
@@ -0,0 +1,411 @@
+type secretValueObject map[string]secretValueVersion
+
+// secretValueVersion holds all the metadata for a specific version
+// of a secret
+type secretValueVersion struct {
+	Created   time.Time `json:"created"`
+	CreatedBy string    `json:"created_by"`
+	Version   int       `json:"version"`
+	Value     string    `json:"value"`
+}
+
+// ensure SecretsManagerStore confirms to Store interface
+var _ Store = &SecretsManagerStore{}
+
+// SecretsManagerStore implements the Store interface for storing secrets in SSM Parameter
+// Store
+type SecretsManagerStore struct {
+	svc    secretsmanageriface.SecretsManagerAPI
+	stsSvc stsiface.STSAPI
+}
+
+// NewSecretsManagerStore creates a new SecretsManagerStore
+func NewSecretsManagerStore(numRetries int) (*SecretsManagerStore, error) {
+	session, region, err := getSession(numRetries)
+	if err != nil {
+		return nil, err
+	}
+
+	svc := secretsmanager.New(session, &aws.Config{
+		MaxRetries: aws.Int(numRetries),
+		Region:     region,
+	})
+
+	stsSvc := sts.New(session, &aws.Config{
+		MaxRetries: aws.Int(numRetries),
+		Region:     region,
+	})
+
+	return &SecretsManagerStore{
+		svc:    svc,
+		stsSvc: stsSvc,
+	}, nil
+}
+
+// Write writes a given value to a secret identified by id.  If the secret
+// already exists, then write a new version.
+func (s *SecretsManagerStore) Write(id SecretId, value string) error {
+	version := 1
+	// first read to get the current version
+	latest, err := s.readLatest(id.Service)
+	mustCreate := false
+
+	if err != nil {
+		if len(value) == 0 {
+			return err
+		}
+		if err != ErrSecretNotFound {
+			if awsErr, ok := err.(awserr.Error); ok {
+				if awsErr.Code() == secretsmanager.ErrCodeResourceNotFoundException {
+					mustCreate = true
+				} else {
+					return err
+				}
+			} else {
+				return err
+			}
+		}
+	}
+
+	if len(value) == 0 {
+		if mustCreate {
+			return err
+		}
+		if _, ok := latest[id.Key]; ok {
+			delete(latest, id.Key)
+		} else {
+			return ErrSecretNotFound
+		}
+	} else {
+		if prop, ok := latest[id.Key]; ok {
+			version = prop.Version + 1
+		}
+
+		user, err := s.getCurrentUser()
+		if err != nil {
+			return err
+		}
+
+		latest[id.Key] = secretValueVersion{
+			Version:   version,
+			Value:     value,
+			Created:   time.Now().UTC(),
+			CreatedBy: user,
+		}
+	}
+
+	contents, err := json.Marshal(latest)
+	if err != nil {
+		return err
+	}
+
+	if mustCreate {
+		createSecretValueInput := &secretsmanager.CreateSecretInput{
+			Name:         aws.String(id.Service),
+			SecretString: aws.String(string(contents)),
+		}
+		_, err = s.svc.CreateSecret(createSecretValueInput)
+		if err != nil {
+			return err
+		}
+	} else {
+		putSecretValueInput := &secretsmanager.PutSecretValueInput{
+			SecretId:      aws.String(id.Service),
+			SecretString:  aws.String(string(contents)),
+			VersionStages: []*string{aws.String("AWSCURRENT"), aws.String("CHAMBER" + string(version))},
+		}
+		_, err = s.svc.PutSecretValue(putSecretValueInput)
```

Comment:
I'm curious if we should prevent updates to secrets that have `RotationEnabled=true` to prevent conflicts with secrets being managed by external Lambdas? Also, with regards to versions and stages, what's the rationale of introducing a `CHAMBER+version` stage? I'm wondering if it would make sense to just leave the Chamber label out to prevent complications in the future should Chamber ever want to utilize this for a rotation strategy of its own.

Reply:
Also, from the docs on `PutSecretValue`:
> If another version of this secret already exists, then this operation does not automatically move any staging labels other than those that you explicitly specify in the `VersionStages` parameter

I'm wondering if we should just omit `VersionStages` and let AWS set these automatically rather than introduce potential conflicts or errors?

Reply:
> I'm curious if we should prevent updates to secrets that have RotationEnabled=true to prevent conflicts with secrets being managed by external Lambdas? 

@eculver I'm not sure any design that supports Chamber functionality could reasonably interop with the AWS-provided Lambdas for rotating secrets. To use Secrets Manager as a Chamber store, we need to manually manage the secret's metadata together with the secret's name and value. None of the AWS-provided Lambdas for rotating secrets are going to be aware of that metadata management (and honestly, just the structure of the secret value is probably a barrier to using those AWS-provided Lambdas for rotation). Instead, if a user wishes to use the Secrets Manager store and enable automatic rotation, they'll need to use Chamber in their rotation Lambda -- meaning we wouldn't be able to prevent updates to secrets that have `RotationEnabled=true` or those rotation Lambdas won't work. 🌀

Reply:
> what's the rationale of introducing a CHAMBER+version stage

This allows us to keep more than 2 versions of a secret. Without an additional staging label, all but the `AWSCURRENT` and `AWSPREVIOUS` versions of the secret are considered "deprecated" by Secrets Manager and subject to deletion. In my experience, these get deleted pretty darn quickly.

Reply:
> I'm not sure any design that supports Chamber functionality could reasonably interop with the AWS-provided Lambdas for rotating secrets. 

This is exactly my point. Chamber can technically still operate on the secrets that are managed by Lambdas which would cause problems. I'm thinking that a simple check for if `RotationEnabled=true` to ensure that the lack of interoperability doesn't cause problems.

This is my only concern at this point. Everything else is looking great.

Reply:
> This is exactly my point. Chamber can technically still operate on the secrets that are managed by Lambdas which would cause problems. I'm thinking that a simple check for if `RotationEnabled=true` to ensure that the lack of interoperability doesn't cause problems.

@eculver I think I understand, but let me make sure. You're saying that if `RotationEnabled=true` we can _read_ the secrets, but we may not **_write_** the secrets, correct? (And this would only be applicable if the secret already exists, of course.) This approach makes sense to me now that the data structure is read-interoperable.

One tricky bit is that we don't know value of `RotationEnabled` at this point in the code. 🤔

Reply:
> I think I understand, but let me make sure. You're saying that if `RotationEnabled=true` we can _read_ the secrets, but we may not **_write_** the secrets, correct? (And this would only be applicable if the secret already exists, of course.) This approach makes sense to me now that the data structure is read-interoperable.

@eculver I went ahead and implemented this via 583cf39